Lib/test/test_peepholer.py

```diff
@@ -1071,7 +1071,17 @@ def test_no_unsafe_static_swap(self):
             ('POP_TOP', 0, 4),
             ('RETURN_VALUE', 5)
         ]
-        self.cfg_optimization_test(insts, insts, consts=list(range(3)), nlocals=1)
+        expected_insts = [
+            ('LOAD_CONST', 0, 1),
+            ('LOAD_CONST', 1, 2),
+            ('LOAD_CONST', 2, 3),
+            ('SWAP', 3, 4),
```

Comment:
I was surprised that this `SWAP` is still here, since allowing `apply_static_swaps` to eliminate it is the main point of this optimization. But on second thought it makes sense; `apply_static_swaps` reads instructions forward from the `SWAP` instruction that triggers it, and that will happen before we reach the `STORE_FAST` instruction that we will replace with `POP_TOP`. So `apply_static_swaps` isn't currently able to benefit from replacing the `STORE_FAST` with `POP_TOP`.

Short of making `optimize_basic_block` multi-pass, I think to address this we'd need to effectively duplicate the `STORE_FAST` -> `POP_TOP` optimization inside `apply_static_swaps`. Which isn't that bad; it's not a lot of code. That could also be a separate PR, though.

Reply:
How about adding apply_static_swap as the last phase of optimize_basic_block?
From the single responsibility principle, IIUC will be more convenient to implement. But it will cost more optimization time, but I am not sure it will be acceptable. 

Maybe @iritkatriel can comment on this point.


````patch

--- a/Python/flowgraph.c
+++ b/Python/flowgraph.c
@@ -1528,10 +1528,6 @@ optimize_basic_block(PyObject *const_cache, basicblock *bb, PyObject *consts)
                     INSTR_SET_OP0(inst, NOP);
                     break;
                 }
-                if (swaptimize(bb, &i) < 0) {
-                    goto error;
-                }
-                apply_static_swaps(bb, i);
                 break;
             case KW_NAMES:
                 break;
@@ -1546,6 +1542,16 @@ optimize_basic_block(PyObject *const_cache, basicblock *bb, PyObject *consts)
                 assert (!HAS_CONST(inst->i_opcode));
         }
     }
+
+    for (int i = 0; i < bb->b_iused; i++) {
+        cfg_instr *inst = &bb->b_instr[i];
+        if (inst->i_opcode == SWAP) {
+            if (swaptimize(bb, &i) < 0) {
+                    goto error;
+            }
+        }
+        apply_static_swaps(bb, i);
+    }
     return SUCCESS;
 error:
     return ERROR;
````


## AS-IS
````

>>> import dis
>>> def f(x, y):
...     a, a = x, y
...     return a
...
>>> dis.dis(f)
  1           0 RESUME                   0

  2           2 LOAD_FAST                0 (x)
              4 LOAD_FAST                1 (y)
              6 SWAP                     2
              8 POP_TOP
             10 STORE_FAST               2 (a)

  3          12 LOAD_FAST                2 (a)
             14 RETURN_VALUE
````

## TO-BE
````
>>> import dis
>>> def f(x, y):
...     a, a = x, y
...     return a
...
>>> dis.dis(f)
  1           0 RESUME                   0

  2           2 LOAD_FAST                0 (x)
              4 LOAD_FAST                1 (y)
              6 STORE_FAST               2 (a)
              8 POP_TOP

  3          10 LOAD_FAST                2 (a)
             12 RETURN_VALUE
```

Reply:
Optimization time is not currently a concern AFAIK.

Reply:
I don't see any other optimizations in `optimize_basic_block` that would depend on having the `SWAP` optimizations run first, so the idea of moving the `SWAP` optimization to a separate second loop over the block seems reasonable enough.